File: go/mergesort1/mergesort1.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"strconv"
	"strings"
	"flag"
	//"time"
)

var (
	memprofile = flag.String("memprofile", "", "write memory profile to this file")
	cpuprofile = flag.String("cpuprofile", "", "write cpu profile to this file")
	infile     = flag.String("infile", "../data/array/array.in", "Matrices to multiply")
)

func merge(arr []int64, left, mid, right int) {
	subArrayOne := mid - left + 1
	subArrayTwo := right - mid

	leftArray := make([]int64, subArrayOne)
	rightArray := make([]int64, subArrayTwo)

	for i := 0; i < subArrayOne; i++ {
		leftArray[i] = arr[left+i]
	}
	for j := 0; j < subArrayTwo; j++ {
		rightArray[j] = arr[mid+1+j]
	}

	indexOfSubArrayOne := 0
	indexOfSubArrayTwo := 0
	indexOfMergedArray := left

	for indexOfSubArrayOne < subArrayOne && indexOfSubArrayTwo < subArrayTwo {
		if leftArray[indexOfSubArrayOne] <= rightArray[indexOfSubArrayTwo] {
			arr[indexOfMergedArray] = leftArray[indexOfSubArrayOne]
			indexOfSubArrayOne++
		} else {
			arr[indexOfMergedArray] = rightArray[indexOfSubArrayTwo]
			indexOfSubArrayTwo++
		}
		indexOfMergedArray++
	}

	for indexOfSubArrayOne < subArrayOne {
		arr[indexOfMergedArray] = leftArray[indexOfSubArrayOne]
		indexOfSubArrayOne++
		indexOfMergedArray++
	}

	for indexOfSubArrayTwo < subArrayTwo {
		arr[indexOfMergedArray] = rightArray[indexOfSubArrayTwo]
		indexOfSubArrayTwo++
		indexOfMergedArray++
	}
}

func mergeSort(arr []int64, begin, end int) {
	if begin >= end {
		return
	}

	mid := begin + (end-begin)/2
	mergeSort(arr, begin, mid)
	mergeSort(arr, mid+1, end)
	merge(arr, begin, mid, end)
}

func main() {
	flag.Parse()
	
	file, _ := os.Open(*infile)
	defer file.Close()

	scanner := bufio.NewScanner(file)

	scanner.Scan()
	arr_length := strings.Fields(scanner.Text())

	n, _ := strconv.Atoi(arr_length[0])

	arr := make([]int64, n)
	scanner.Scan()
	line := scanner.Text()
	iss := strings.Fields(line)
	
	for i, numStr := range iss{
		arr[i], _ = strconv.ParseInt(numStr, 10, 64)
	}

	//start := time.Now()
	mergeSort(arr, 0, n-1)
	//end := time.Now()

	for _, element := range arr {
		fmt.Println(element)
	}

	//fmt.Println("Mergesort time for", n, "elements:", end.Sub(start))
}

File: cpp/mergesort1/mergesort1.cc
#include <bits/stdc++.h>
#include <gperftools/profiler.h>
#include <ctime>
using namespace std;

void merge(vector<long long> &array, int const left, int const mid,
           int const right)
{
    long long const subArrayOne = mid - left + 1;
    long long const subArrayTwo = right - mid;
 
    // Create temp arrays
    auto *leftArray = new long long[subArrayOne],
         *rightArray = new long long[subArrayTwo];
 
    // Copy data to temp arrays leftArray[] and rightArray[]
    for (auto i = 0; i < subArrayOne; i++)
        leftArray[i] = array[left + i];
    for (auto j = 0; j < subArrayTwo; j++)
        rightArray[j] = array[mid + 1 + j];
 
    auto indexOfSubArrayOne = 0, indexOfSubArrayTwo = 0;
    int indexOfMergedArray = left;
 
    // Merge the temp arrays back into array[left..right]
    while (indexOfSubArrayOne < subArrayOne
           && indexOfSubArrayTwo < subArrayTwo) {
        if (leftArray[indexOfSubArrayOne]
            <= rightArray[indexOfSubArrayTwo]) {
            array[indexOfMergedArray]
                = leftArray[indexOfSubArrayOne];
            indexOfSubArrayOne++;
        }
        else {
            array[indexOfMergedArray]
                = rightArray[indexOfSubArrayTwo];
            indexOfSubArrayTwo++;
        }
        indexOfMergedArray++;
    }
 
    // Copy the remaining elements of
    // left[], if there are any
    while (indexOfSubArrayOne < subArrayOne) {
        array[indexOfMergedArray]
            = leftArray[indexOfSubArrayOne];
        indexOfSubArrayOne++;
        indexOfMergedArray++;
    }
 
    // Copy the remaining elements of
    // right[], if there are any
    while (indexOfSubArrayTwo < subArrayTwo) {
        array[indexOfMergedArray]
            = rightArray[indexOfSubArrayTwo];
        indexOfSubArrayTwo++;
        indexOfMergedArray++;
    }
    delete[] leftArray;
    delete[] rightArray;
}
 
// begin is for left index and end is right index
// of the sub-array of arr to be sorted
void mergeSort(vector<long long> &array, int const begin, int const end)
{
    if (begin >= end)
        return;
 
    int mid = begin + (end - begin) / 2;
    mergeSort(array, begin, mid);
    mergeSort(array, mid + 1, end);
    merge(array, begin, mid, end);
}

int main(){
    fstream f("../data/array/array.in");
    if (!f.good()){
        std::cerr << "double check file path\n";
        return 1;
    }

    string line;
    getline(f,line);
    istringstream arr_length(line);
    int n;
    arr_length >> n;
    //cout << n << endl;

    long long num;
    getline(f,line);
    std::istringstream iss(line);
    std::vector<long long> arr;
    while (iss >> num){
        arr.push_back(num);
    }
    
    //clock_t start = clock();
    mergeSort(arr,0,n-1);
    //clock_t end = clock();
    
    for (auto &element : arr){
        cout << element << endl;
    }

    //cout<< "Quicksort time for " << n << " elements: "<< (float)(end - start)/ CLOCKS_PER_SEC << " "<<endl;;
    f.close();
    return 0;
}